content/blog/2019-09-25-bril-type-inference.md

```diff
@@ -0,0 +1,204 @@
++++
+title = "Type Inference for Bril"
+extra.bio = """
+  Christopher Roman is a second semester MEng student in computer science. He is interested in compilers and distributed systems. He is also the best [Melee](https://en.wikipedia.org/wiki/Super_Smash_Bros._Melee) player at Cornell :)
+"""
+[[extra.authors]]
+name = "Christopher Roman"
++++
+
+## Overview
+The goal of this is to build a type inferrer for Bril. Ultimately, we want to
+take a Bril program that has some (or none) of the types specified on variables
+and produce a new Bril program that has all the correct type annotations.
+For example, this program:
+```
+main {
+  v0 = const 1;
+  v1 = const 2;
+  v2 = add v0 v1;
+  print v2;
+}
+```
+should be transformed into the program:
+```
+main {
+  v0: int = const 1;
+  v1: int = const 2;
+  v2: int = add v0 v1;
+  print v2;
+}
+```
+This is useful in cases where the frontend isn't typed. For example, someone
+building a Python frontend would need to specify the types of variables, which
+may be annoying. This takes some burden off the frontend implementation.
+Additionally, it makes it quicker to write Bril programs because we don't have
+to worry about correctly typing the variables.
```

Comment:
FWIW, I think the Python motivation is somewhat imperfect—for arbitrary Python programs, it may be *impossible* to come up with a static typing with the same semantics. That's just the nature of a dynamic language. So we'd still be stuck compiling a statically-typed *subset* of Python, even if the backend infers the types.

But making it easier to write the IR makes sense!

Reply:
I see, I never thought about that. To give some clearer motivation, I've just removed this example in 10c884b